src/sorted_container_iteration.jl

```diff
@@ -0,0 +1,1167 @@
+## Functions in this file implement many kinds of iterations over
+## SortedDict, SortedMultiDict, and SortedSet.
+## The iteration state is a "semitoken".
+## A "token" is a (container,index_into_container) 2-tuple,
+## while semitokens are the second part, index_into_container.
+
+# From tokens.jl:
+# abstract type AbstractSemiToken end
+
+# struct IntSemiToken <: AbstractSemiToken
+#     address::Int
+# end
+
+const SDMContainer = Union{SortedDict, SortedMultiDict}
+const SortedContainer = Union{SDMContainer, SortedSet}
+const Token = Tuple{SortedContainer, IntSemiToken}
```

Comment:
Are you asking why `Token` is implemented as a tuple rather than a `struct`?   This is a legacy issue.  In pre-1.0 Julia, any struct that contained an immutable object had itself to be heap allocated.  However, in pre-1.0 Julia, operations on tuples containing an immutable object didn't always allocate.  In particular, `deref((mysorteddict, semitok))` did not heap-allocate its 2-tuple argument, whereas a hypothetical `deref(Token(mysorteddict, semitok))`, where `Token` is the constructor of a hypothetical struct, would have allocated.

Reply:
More my surprise was that a token included both the container and the indexing object.
I would have expected the token to just be the indexing object (but that is defined as a semitoken)
is the idea that the token is enough to fully retrieve the stored object?
But semitoken is only half of what is expected?

Reply:
I think full tokens are useful, e.g., `deref(token_firstindex(s))` as opposed to `deref((s,firstindex(s)))` where the user needs to be consistent with the argument `s`.  So if it is OK with you, I would like to leave these.